metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/storage/ViewStateStorage.scala

```diff
@@ -0,0 +1,148 @@
+package io.iohk.metronome.hotstuff.service.storage
+
+import cats.implicits._
+import io.iohk.metronome.hotstuff.consensus.ViewNumber
+import io.iohk.metronome.hotstuff.consensus.basic.{
+  Agreement,
+  QuorumCertificate,
+  Phase
+}
+import io.iohk.metronome.storage.{KVStore, KVStoreRead}
+import scodec.{Codec, Encoder, Decoder}
+
+class ViewStateStorage[N, A <: Agreement] private (
+    namespace: N
+)(implicit
+    keys: ViewStateStorage.Keys[A],
+    kvn: KVStore.Ops[N],
+    kvrn: KVStoreRead.Ops[N],
+    codecVN: Codec[ViewNumber],
+    codecQC: Codec[QuorumCertificate[A]],
+    codecH: Codec[A#Hash]
+) {
+  import keys.Key
+
+  private def put[V: Encoder](key: Key[V], value: V) =
+    KVStore[N].put[Key[V], V](namespace, key, value)
+
+  private def read[V: Decoder](key: Key[V]): KVStoreRead[N, V] =
+    KVStoreRead[N].read[Key[V], V](namespace, key).map(_.get)
+
+  def setViewNumber(viewNumber: ViewNumber): KVStore[N, Unit] =
+    put(Key.ViewNumber, viewNumber)
+
+  def setQuorumCertificate(qc: QuorumCertificate[A]): KVStore[N, Unit] =
+    qc.phase match {
+      case Phase.Prepare =>
+        put(Key.PrepareQC, qc)
+      case Phase.PreCommit =>
+        put(Key.LockedQC, qc)
+      case Phase.Commit =>
+        put(Key.CommitQC, qc)
+    }
+
+  def setLastExecutedBlockHash(blockHash: A#Hash): KVStore[N, Unit] =
+    put(Key.LastExecutedBlockHash, blockHash)
+
+  def getBundle: KVStoreRead[N, ViewStateStorage.Bundle[A]] =
+    (
+      read(Key.ViewNumber),
+      read(Key.PrepareQC),
+      read(Key.LockedQC),
+      read(Key.CommitQC),
+      read(Key.LastExecutedBlockHash)
+    ).mapN(ViewStateStorage.Bundle.apply[A] _)
+
+}
+
+object ViewStateStorage {
+
+  /** Storing elements of the view state individually under separate keys,
+    * because they get written independently.
+    */
+  trait Keys[A <: Agreement] {
+    sealed trait Key[V]
+    object Key {
+      case object ViewNumber            extends Key[ViewNumber]
+      case object PrepareQC             extends Key[QuorumCertificate[A]]
+      case object LockedQC              extends Key[QuorumCertificate[A]]
+      case object CommitQC              extends Key[QuorumCertificate[A]]
+      case object LastExecutedBlockHash extends Key[A#Hash]
+
+      implicit def encoder[V]: Encoder[Key[V]] =
+        scodec.codecs.implicits.implicitStringCodec
+          .contramap[Key[V]](_.toString)
```

Comment:
Aren't we concerned with storage efficiency? How about using something like `sealed abstract class Key[A <: Agreement](val code: Int)` and using that `code` as storage key?

Reply:
I don't know, there's only a couple of keys. Can't it handle some strings?

I'm concerned with one of @ntallar comments in the other project that reinserting the same records over and over caused RocksDB to grow until it was restarted, like, space after deleted records were not reclaimed automatically.

Reply:
There must be some sort of concern to cause people to use [single letter namespaces](https://github.com/input-output-hk/mantis/blob/develop/src/main/scala/io/iohk/ethereum/db/storage/Namespaces.scala#L4-L14), which got really ugly in the other project where we ran out of intuitive letters and started going down the alphabet. Numbers would be much better in that sense, but I thought this even more foolproof and lazy.

Reply:
But you may be right, why not. Pushed a change to use 1 byte.